File: packages/go/hushspec/conditions_test.go
package hushspec

import (
	"testing"
)

func ctxWithEnv(env string) *RuntimeContext {
	return &RuntimeContext{Environment: env}
}

func ctxWithTimeStr(t string) *RuntimeContext {
	return &RuntimeContext{CurrentTime: t}
}

func ctxWithUserRole(role string) *RuntimeContext {
	return &RuntimeContext{
		User: map[string]interface{}{"role": role},
	}
}

func makeEgressSpecForCond() *HushSpec {
	return &HushSpec{
		HushSpecVersion: "0.1.0",
		Name:            "conditional-test",
		Rules: &Rules{
			Egress: &EgressRule{
				Enabled: true,
				Allow:   []string{"api.openai.com"},
				Default: DefaultActionBlock,
			},
		},
	}
}

func TestContextConditionMatchesEnvironment(t *testing.T) {
	cond := &Condition{
		Context: map[string]interface{}{"environment": "production"},
	}
	if !EvaluateCondition(cond, ctxWithEnv("production")) {
		t.Error("expected condition to match production environment")
	}
}

func TestContextConditionRejectsMismatch(t *testing.T) {
	cond := &Condition{
		Context: map[string]interface{}{"environment": "production"},
	}
	if EvaluateCondition(cond, ctxWithEnv("staging")) {
		t.Error("expected condition to reject staging environment")
	}
}

func TestContextConditionMissingFieldFailsClosed(t *testing.T) {
	cond := &Condition{
		Context: map[string]interface{}{"user.role": "admin"},
	}
	if EvaluateCondition(cond, &RuntimeContext{}) {
		t.Error("expected missing field to fail closed")
	}
}

func TestContextConditionMatchesUserRole(t *testing.T) {
	cond := &Condition{
		Context: map[string]interface{}{"user.role": "admin"},
	}
	if !EvaluateCondition(cond, ctxWithUserRole("admin")) {
		t.Error("expected admin to match")
	}
	if EvaluateCondition(cond, ctxWithUserRole("viewer")) {
		t.Error("expected viewer to not match")
	}
}

func TestContextConditionArrayOrMatch(t *testing.T) {
	cond := &Condition{
		Context: map[string]interface{}{
			"environment": []interface{}{"production", "staging"},
		},
	}
	if !EvaluateCondition(cond, ctxWithEnv("production")) {
		t.Error("expected production to match")
	}
	if !EvaluateCondition(cond, ctxWithEnv("staging")) {
		t.Error("expected staging to match")
	}
	if EvaluateCondition(cond, ctxWithEnv("development")) {
		t.Error("expected development to not match")
	}
}

func TestContextConditionScalarVsArrayMembership(t *testing.T) {
	ctx := &RuntimeContext{
		User: map[string]interface{}{
			"groups": []interface{}{"engineering", "ml-team"},
		},
	}
	cond := &Condition{
		Context: map[string]interface{}{"user.groups": "ml-team"},
	}
	if !EvaluateCondition(cond, ctx) {
		t.Error("expected ml-team to be found in groups array")
	}
}

func TestTimeWindowMatchesDuringBusinessHours(t *testing.T) {
	ctx := ctxWithTimeStr("2026-01-14T10:30:00Z")
	cond := &Condition{
		TimeWindow: &TimeWindowCondition{
			Start:    "09:00",
			End:      "17:00",
			Timezone: "UTC",
		},
	}
	if !EvaluateCondition(cond, ctx) {
		t.Error("expected business hours to match")
	}
}

func TestTimeWindowRejectsOutsideHours(t *testing.T) {
	ctx := ctxWithTimeStr("2026-01-14T20:00:00Z")
	cond := &Condition{
		TimeWindow: &TimeWindowCondition{
			Start:    "09:00",
			End:      "17:00",
			Timezone: "UTC",
		},
	}
	if EvaluateCondition(cond, ctx) {
		t.Error("expected outside hours to not match")
	}
}

func TestTimeWindowDayFilter(t *testing.T) {
	// 2026-01-14 is a Wednesday
	ctx := ctxWithTimeStr("2026-01-14T10:00:00Z")

	condWeekday := &Condition{
		TimeWindow: &TimeWindowCondition{
			Start:    "09:00",
			End:      "17:00",
			Timezone: "UTC",
			Days:     []string{"mon", "tue", "wed", "thu", "fri"},
		},
	}
	if !EvaluateCondition(condWeekday, ctx) {
		t.Error("expected weekday match on Wednesday")
	}

	condWeekend := &Condition{
		TimeWindow: &TimeWindowCondition{
			Start:    "09:00",
			End:      "17:00",
			Timezone: "UTC",
			Days:     []string{"sat", "sun"},
		},
	}
	if EvaluateCondition(condWeekend, ctx) {
		t.Error("expected weekend to not match on Wednesday")
	}
}

func TestTimeWindowWrapsMidnight(t *testing.T) {
	cond := &Condition{
		TimeWindow: &TimeWindowCondition{
			Start:    "22:00",
			End:      "06:00",
			Timezone: "UTC",
		},
	}

	if !EvaluateCondition(cond, ctxWithTimeStr("2026-01-14T23:00:00Z")) {
		t.Error("expected 23:00 to match night window")
	}
	if !EvaluateCondition(cond, ctxWithTimeStr("2026-01-14T03:00:00Z")) {
		t.Error("expected 03:00 to match night window")
	}
	if EvaluateCondition(cond, ctxWithTimeStr("2026-01-14T10:00:00Z")) {
		t.Error("expected 10:00 to not match night window")
	}
}

func TestTimeWindowSameStartEndMeansAllDay(t *testing.T) {
	cond := &Condition{
		TimeWindow: &TimeWindowCondition{
			Start:    "12:00",
			End:      "12:00",
			Timezone: "UTC",
		},
	}
	if !EvaluateCondition(cond, ctxWithTimeStr("2026-01-14T03:00:00Z")) {
		t.Error("expected same start/end to match all day")
	}
}

func TestTimeWindowSupportsMinuteOffsets(t *testing.T) {
	cond := &Condition{
		TimeWindow: &TimeWindowCondition{
			Start:    "05:30",
			End:      "06:30",
			Timezone: "+05:30",
		},
	}
	if !EvaluateCondition(cond, ctxWithTimeStr("2026-01-14T00:15:00Z")) {
		t.Error("expected +05:30 offset to match inside the window")
	}
	if EvaluateCondition(cond, ctxWithTimeStr("2026-01-14T01:15:00Z")) {
		t.Error("expected +05:30 offset to reject outside the window")
	}
}

func TestTimeWindowUsesDSTForIANATimezones(t *testing.T) {
	cond := &Condition{
		TimeWindow: &TimeWindowCondition{
			Start:    "08:30",
			End:      "09:30",
			Timezone: "America/New_York",
		},
	}
	if !EvaluateCondition(cond, ctxWithTimeStr("2026-01-14T13:45:00Z")) {
		t.Error("expected winter New York time to match")
	}
	if !EvaluateCondition(cond, ctxWithTimeStr("2026-07-14T12:45:00Z")) {
		t.Error("expected summer New York time to match under DST")
	}
}

func TestTimeWindowWrapsMidnightWithDayFilter(t *testing.T) {
	cond := &Condition{
		TimeWindow: &TimeWindowCondition{
			Start:    "22:00",
			End:      "06:00",
			Timezone: "UTC",
			Days:     []string{"fri"},
		},
	}
	if !EvaluateCondition(cond, ctxWithTimeStr("2026-01-17T03:00:00Z")) {
		t.Error("expected Saturday early-morning time to count as Friday night")
	}
}

func TestTimeWindowInvalidTimezoneFailsClosed(t *testing.T) {
	cond := &Condition{
		TimeWindow: &TimeWindowCondition{
			Start:    "09:00",
			End:      "17:00",
			Timezone: "America/NeYork",
		},
	}
	if EvaluateCondition(cond, ctxWithTimeStr("2026-01-14T13:30:00Z")) {
		t.Error("expected invalid timezone to fail closed")
	}
}

func TestAllOfRequiresAllConditions(t *testing.T) {
	cond := &Condition{
		AllOf: []Condition{
			{Context: map[string]interface{}{"environment": "production"}},
			{Context: map[string]interface{}{"user.role": "admin"}},
		},
	}

	fullCtx := &RuntimeContext{
		Environment: "production",
		User:        map[string]interface{}{"role": "admin"},
	}
	if !EvaluateCondition(cond, fullCtx) {
		t.Error("expected both conditions to match")
	}

	if EvaluateCondition(cond, ctxWithEnv("production")) {
		t.Error("expected partial match to fail")
	}
}

func TestAnyOfRequiresAnyCondition(t *testing.T) {
	cond := &Condition{
		AnyOf: []Condition{
			{Context: map[string]interface{}{"environment": "production"}},
			{Context: map[string]interface{}{"environment": "staging"}},
		},
	}

	if !EvaluateCondition(cond, ctxWithEnv("production")) {
		t.Error("expected production to match")
	}
	if !EvaluateCondition(cond, ctxWithEnv("staging")) {
		t.Error("expected staging to match")
	}
	if EvaluateCondition(cond, ctxWithEnv("development")) {
		t.Error("expected development to not match")
	}
}

func TestNotNegatesCondition(t *testing.T) {
	cond := &Condition{
		Not: &Condition{
			Context: map[string]interface{}{"environment": "production"},
		},
	}

	if EvaluateCondition(cond, ctxWithEnv("production")) {
		t.Error("expected NOT production to fail")
	}
	if !EvaluateCondition(cond, ctxWithEnv("staging")) {
		t.Error("expected NOT production to pass for staging")
	}
}

func TestNestedCompoundConditions(t *testing.T) {
	cond := &Condition{
		AllOf: []Condition{
			{
				TimeWindow: &TimeWindowCondition{
					Start:    "09:00",
					End:      "17:00",
					Timezone: "UTC",
				},
			},
			{Context: map[string]interface{}{"environment": "production"}},
			{
				AnyOf: []Condition{
					{Context: map[string]interface{}{"user.role": "admin"}},
					{Context: map[string]interface{}{"user.role": "sre"}},
				},
			},
		},
	}

	ctx := &RuntimeContext{
		Environment: "production",
		CurrentTime: "2026-01-14T10:00:00Z",
		User:        map[string]interface{}{"role": "admin"},
	}
	if !EvaluateCondition(cond, ctx) {
		t.Error("expected nested compound to match")
	}

	ctxViewer := &RuntimeContext{
		Environment: "production",
		CurrentTime: "2026-01-14T10:00:00Z",
		User:        map[string]interface{}{"role": "viewer"},
	}
	if EvaluateCondition(cond, ctxViewer) {
		t.Error("expected viewer to fail nested compound")
	}
}

func TestEmptyConditionAlwaysTrue(t *testing.T) {
	cond := &Condition{}
	if !EvaluateCondition(cond, &RuntimeContext{}) {
		t.Error("expected empty condition to be true")
	}
}

func TestMaxNestingDepthExceeded(t *testing.T) {
	cond := &Condition{
		Context: map[string]interface{}{"environment": "production"},
	}
	for i := 0; i < 12; i++ {
		cond = &Condition{AllOf: []Condition{*cond}}
	}
	if EvaluateCondition(cond, ctxWithEnv("production")) {
		t.Error("expected max nesting to fail")
	}
}

func TestEvaluateWithContextPassesWhenConditionMet(t *testing.T) {
	spec := makeEgressSpecForCond()
	action := &EvaluationAction{Type: "egress", Target: "api.openai.com"}
	ctx := &RuntimeContext{Environment: "production"}
	conditions := map[string]*Condition{
		"egress": {Context: map[string]interface{}{"environment": "production"}},
	}

	result := EvaluateWithContext(spec, action, ctx, conditions)
	if result.Decision != DecisionAllow {
		t.Errorf("expected allow, got %s", result.Decision)
	}
}

func TestEvaluateWithContextSkipsRuleWhenConditionFails(t *testing.T) {
	spec := makeEgressSpecForCond()
	action := &EvaluationAction{Type: "egress", Target: "evil.example.com"}
	ctx := &RuntimeContext{Environment: "staging"}
	conditions := map[string]*Condition{
		"egress": {Context: map[string]interface{}{"environment": "production"}},
	}

	result := EvaluateWithContext(spec, action, ctx, conditions)
	if result.Decision != DecisionAllow {
		t.Errorf("expected allow (rule disabled), got %s", result.Decision)
	}
}

func TestEvaluateWithContextEnforcesRuleWhenConditionMet(t *testing.T) {
	spec := makeEgressSpecForCond()
	action := &EvaluationAction{Type: "egress", Target: "evil.example.com"}
	ctx := &RuntimeContext{Environment: "production"}
	conditions := map[string]*Condition{
		"egress": {Context: map[string]interface{}{"environment": "production"}},
	}

	result := EvaluateWithContext(spec, action, ctx, conditions)
	if result.Decision != DecisionDeny {
		t.Errorf("expected deny, got %s", result.Decision)
	}
}

func TestEvaluateWithContextNoConditionsBehavesLikeEvaluate(t *testing.T) {
	spec := makeEgressSpecForCond()
	action := &EvaluationAction{Type: "egress", Target: "evil.example.com"}
	ctx := &RuntimeContext{}
	conditions := map[string]*Condition{}

	result := EvaluateWithContext(spec, action, ctx, conditions)
	if result.Decision != DecisionDeny {
		t.Errorf("expected deny, got %s", result.Decision)
	}
}

func TestEvaluateWithContextMissingContextFailsClosed(t *testing.T) {
	spec := makeEgressSpecForCond()
	action := &EvaluationAction{Type: "egress", Target: "api.openai.com"}
	ctx := &RuntimeContext{}
	conditions := map[string]*Condition{
		"egress": {Context: map[string]interface{}{"environment": "production"}},
	}

	result := EvaluateWithContext(spec, action, ctx, conditions)
	if result.Decision != DecisionAllow {
		t.Errorf("expected allow (condition fails, rule disabled), got %s", result.Decision)
	}
}

func TestEvaluateWithContextCompoundCondition(t *testing.T) {
	spec := makeEgressSpecForCond()
	action := &EvaluationAction{Type: "egress", Target: "evil.example.com"}
	conditions := map[string]*Condition{
		"egress": {
			AllOf: []Condition{
				{Context: map[string]interface{}{"environment": "production"}},
				{Context: map[string]interface{}{"user.role": "admin"}},
			},
		},
	}

	fullCtx := &RuntimeContext{
		Environment: "production",
		User:        map[string]interface{}{"role": "admin"},
	}
	result := EvaluateWithContext(spec, action, fullCtx, conditions)
	if result.Decision != DecisionDeny {
		t.Errorf("expected deny, got %s", result.Decision)
	}

	partialCtx := &RuntimeContext{Environment: "production"}
	result2 := EvaluateWithContext(spec, action, partialCtx, conditions)
	if result2.Decision != DecisionAllow {
		t.Errorf("expected allow (partial condition fails), got %s", result2.Decision)
	}
}

File: packages/hushspec/tests/conditions.test.ts
import { describe, it, expect } from 'vitest';
import {
  evaluateCondition,
  evaluateWithContext,
  type Condition,
  type RuntimeContext,
} from '../src/conditions.js';
import type { HushSpec } from '../src/schema.js';

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function ctxWithEnv(env: string): RuntimeContext {
  return { environment: env };
}

function ctxWithTime(time: string): RuntimeContext {
  return { current_time: time };
}

function ctxWithUserRole(role: string): RuntimeContext {
  return { user: { role } };
}

function makeEgressSpec(): HushSpec {
  return {
    hushspec: '0.1.0',
    name: 'conditional-test',
    rules: {
      egress: {
        enabled: true,
        allow: ['api.openai.com'],
        default: 'block',
      },
    },
  };
}

function makeToolAccessSpec(): HushSpec {
  return {
    hushspec: '0.1.0',
    name: 'conditional-tool-test',
    rules: {
      tool_access: {
        enabled: true,
        allow: ['deploy'],
        block: ['danger_tool'],
        default: 'block',
      },
    },
  };
}

// ---------------------------------------------------------------------------
// Context conditions
// ---------------------------------------------------------------------------

describe('evaluateCondition', () => {
  describe('context conditions', () => {
    it('matches environment', () => {
      const cond: Condition = {
        context: { environment: 'production' },
      };
      expect(evaluateCondition(cond, ctxWithEnv('production'))).toBe(true);
    });

    it('rejects mismatch', () => {
      const cond: Condition = {
        context: { environment: 'production' },
      };
      expect(evaluateCondition(cond, ctxWithEnv('staging'))).toBe(false);
    });

    it('missing context field fails closed', () => {
      const cond: Condition = {
        context: { 'user.role': 'admin' },
      };
      expect(evaluateCondition(cond, {})).toBe(false);
    });

    it('matches user role', () => {
      const cond: Condition = {
        context: { 'user.role': 'admin' },
      };
      expect(evaluateCondition(cond, ctxWithUserRole('admin'))).toBe(true);
      expect(evaluateCondition(cond, ctxWithUserRole('viewer'))).toBe(false);
    });

    it('array of expected values (OR within a key)', () => {
      const cond: Condition = {
        context: { environment: ['production', 'staging'] },
      };
      expect(evaluateCondition(cond, ctxWithEnv('production'))).toBe(true);
      expect(evaluateCondition(cond, ctxWithEnv('staging'))).toBe(true);
      expect(evaluateCondition(cond, ctxWithEnv('development'))).toBe(false);
    });

    it('scalar expected vs array actual (membership check)', () => {
      const ctx: RuntimeContext = {
        user: { groups: ['engineering', 'ml-team'] },
      };
      const cond: Condition = {
        context: { 'user.groups': 'ml-team' },
      };
      expect(evaluateCondition(cond, ctx)).toBe(true);
    });
  });

  // -----------------------------------------------------------------------
  // Time window conditions
  // -----------------------------------------------------------------------

  describe('time window conditions', () => {
    it('matches during business hours', () => {
      const ctx = ctxWithTime('2026-01-14T10:30:00Z');
      const cond: Condition = {
        time_window: {
          start: '09:00',
          end: '17:00',
          timezone: 'UTC',
        },
      };
      expect(evaluateCondition(cond, ctx)).toBe(true);
    });

    it('rejects outside hours', () => {
      const ctx = ctxWithTime('2026-01-14T20:00:00Z');
      const cond: Condition = {
        time_window: {
          start: '09:00',
          end: '17:00',
          timezone: 'UTC',
        },
      };
      expect(evaluateCondition(cond, ctx)).toBe(false);
    });

    it('filters by day of week', () => {
      // 2026-01-14 is a Wednesday
      const ctx = ctxWithTime('2026-01-14T10:00:00Z');

      const weekdayCond: Condition = {
        time_window: {
          start: '09:00',
          end: '17:00',
          timezone: 'UTC',
          days: ['mon', 'tue', 'wed', 'thu', 'fri'],
        },
      };
      expect(evaluateCondition(weekdayCond, ctx)).toBe(true);

      const weekendCond: Condition = {
        time_window: {
          start: '09:00',
          end: '17:00',
          timezone: 'UTC',
          days: ['sat', 'sun'],
        },
      };
      expect(evaluateCondition(weekendCond, ctx)).toBe(false);
    });

    it('wraps midnight', () => {
      const condNight: Condition = {
        time_window: {
          start: '22:00',
          end: '06:00',
          timezone: 'UTC',
        },
      };
      expect(evaluateCondition(condNight, ctxWithTime('2026-01-14T23:00:00Z'))).toBe(true);
      expect(evaluateCondition(condNight, ctxWithTime('2026-01-14T03:00:00Z'))).toBe(true);
      expect(evaluateCondition(condNight, ctxWithTime('2026-01-14T10:00:00Z'))).toBe(false);
    });

    it('same start and end means all day', () => {
      const cond: Condition = {
        time_window: {
          start: '12:00',
          end: '12:00',
          timezone: 'UTC',
        },
      };
      expect(evaluateCondition(cond, ctxWithTime('2026-01-14T03:00:00Z'))).toBe(true);
    });
  });

  // -----------------------------------------------------------------------
  // Compound conditions
  // -----------------------------------------------------------------------

  describe('compound conditions', () => {
    it('all_of requires all conditions', () => {
      const cond: Condition = {
        all_of: [
          { context: { environment: 'production' } },
          { context: { 'user.role': 'admin' } },
        ],
      };

      const fullCtx: RuntimeContext = {
        environment: 'production',
        user: { role: 'admin' },
      };
      expect(evaluateCondition(cond, fullCtx)).toBe(true);

      // Only env matches
      expect(evaluateCondition(cond, ctxWithEnv('production'))).toBe(false);
    });

    it('any_of requires any condition', () => {
      const cond: Condition = {
        any_of: [
          { context: { environment: 'production' } },
          { context: { environment: 'staging' } },
        ],
      };

      expect(evaluateCondition(cond, ctxWithEnv('production'))).toBe(true);
      expect(evaluateCondition(cond, ctxWithEnv('staging'))).toBe(true);
      expect(evaluateCondition(cond, ctxWithEnv('development'))).toBe(false);
    });

    it('not negates condition', () => {
      const cond: Condition = {
        not: { context: { environment: 'production' } },
      };

      expect(evaluateCondition(cond, ctxWithEnv('production'))).toBe(false);
      expect(evaluateCondition(cond, ctxWithEnv('staging'))).toBe(true);
    });

    it('nested compound conditions', () => {
      // Business hours AND production AND (admin OR sre)
      const cond: Condition = {
        all_of: [
          {
            time_window: {
              start: '09:00',
              end: '17:00',
              timezone: 'UTC',
            },
          },
          { context: { environment: 'production' } },
          {
            any_of: [
              { context: { 'user.role': 'admin' } },
              { context: { 'user.role': 'sre' } },
            ],
          },
        ],
      };

      const ctx: RuntimeContext = {
        environment: 'production',
        current_time: '2026-01-14T10:00:00Z',
        user: { role: 'admin' },
      };
      expect(evaluateCondition(cond, ctx)).toBe(true);

      const ctxViewer: RuntimeContext = {
        ...ctx,
        user: { role: 'viewer' },
      };
      expect(evaluateCondition(cond, ctxViewer)).toBe(false);
    });
  });

  describe('edge cases', () => {
    it('empty condition is always true', () => {
      expect(evaluateCondition({}, {})).toBe(true);
    });

    it('max nesting depth exceeded fails closed', () => {
      // Build deeply nested condition
      let cond: Condition = { context: { environment: 'production' } };
      for (let i = 0; i < 12; i++) {
        cond = { all_of: [cond] };
      }
      expect(evaluateCondition(cond, ctxWithEnv('production'))).toBe(false);
    });
  });
});

// ---------------------------------------------------------------------------
// evaluateWithContext
// ---------------------------------------------------------------------------

describe('evaluateWithContext', () => {
  it('passes when condition is met', () => {
    const spec = makeEgressSpec();
    const action = { type: 'egress', target: 'api.openai.com' };
    const ctx: RuntimeContext = { environment: 'production' };
    const conditions: Record<string, Condition> = {
      egress: { context: { environment: 'production' } },
    };

    const result = evaluateWithContext(spec, action, ctx, conditions);
    expect(result.decision).toBe('allow');
  });

  it('skips rule when condition fails', () => {
    const spec = makeEgressSpec();
    const action = { type: 'egress', target: 'evil.example.com' };
    const ctx: RuntimeContext = { environment: 'staging' };
    const conditions: Record<string, Condition> = {
      egress: { context: { environment: 'production' } },
    };

    const result = evaluateWithContext(spec, action, ctx, conditions);
    expect(result.decision).toBe('allow');
  });

  it('enforces rule when condition is met', () => {
    const spec = makeEgressSpec();
    const action = { type: 'egress', target: 'evil.example.com' };
    const ctx: RuntimeContext = { environment: 'production' };
    const conditions: Record<string, Condition> = {
      egress: { context: { environment: 'production' } },
    };

    const result = evaluateWithContext(spec, action, ctx, conditions);
    expect(result.decision).toBe('deny');
  });

  it('no conditions behaves like evaluate', () => {
    const spec = makeEgressSpec();
    const action = { type: 'egress', target: 'evil.example.com' };
    const ctx: RuntimeContext = {};
    const conditions: Record<string, Condition> = {};

    const result = evaluateWithContext(spec, action, ctx, conditions);
    expect(result.decision).toBe('deny');
  });

  it('tool access with time window condition', () => {
    const spec = makeToolAccessSpec();
    const action = { type: 'tool_call', target: 'deploy' };
    const conditions: Record<string, Condition> = {
      tool_access: {
        time_window: {
          start: '09:00',
          end: '17:00',
          timezone: 'UTC',
        },
      },
    };

    const ctxInside: RuntimeContext = { current_time: '2026-01-14T10:00:00Z' };
    const resultInside = evaluateWithContext(spec, action, ctxInside, conditions);
    expect(resultInside.decision).toBe('allow');

    const ctxOutside: RuntimeContext = { current_time: '2026-01-14T20:00:00Z' };
    const resultOutside = evaluateWithContext(spec, action, ctxOutside, conditions);
    expect(resultOutside.decision).toBe('allow');
  });

  it('missing context fails closed', () => {
    const spec = makeEgressSpec();
    const action = { type: 'egress', target: 'api.openai.com' };
    const ctx: RuntimeContext = {};
    const conditions: Record<string, Condition> = {
      egress: { context: { environment: 'production' } },
    };

    const result = evaluateWithContext(spec, action, ctx, conditions);
    expect(result.decision).toBe('allow');
  });

  it('compound condition', () => {
    const spec = makeEgressSpec();
    const action = { type: 'egress', target: 'evil.example.com' };
    const conditions: Record<string, Condition> = {
      egress: {
        all_of: [
          { context: { environment: 'production' } },
          { context: { 'user.role': 'admin' } },
        ],
      },
    };

    const fullCtx: RuntimeContext = {
      environment: 'production',
      user: { role: 'admin' },
    };
    const result = evaluateWithContext(spec, action, fullCtx, conditions);
    expect(result.decision).toBe('deny');

    const partialCtx: RuntimeContext = { environment: 'production' };
    const result2 = evaluateWithContext(spec, action, partialCtx, conditions);
    expect(result2.decision).toBe('allow');
  });
});
